### PR TITLE
Improve Quoting (now with webhooks!)

### PR DESCRIPTION
**Pull request by zhangchuck**
# Description

Continuing to refactor/improve quoting (#21)

# What was done

- `requirements.txt` updated to take the latest version of discord.py rewrite (#28)
- adopted the `async/await` syntax (native co-routines) instead of `@asyncio/yield from) (related to #28)
- implemented quoting via webhooks (#35)
     - if the bot does not have 'manage webhooks' permissions, it will revert to using the old way of sending messages
    - if there is an existing webhook, the bot will make use of it
    - if there is not an existing webhook, the bot will create one
    - see the function `_get_hook`
- change how time is displayed on quotes (#18)
    - we no longer quote the absolute time, we only quote relative time
    - times are linked to the original message, so you can still look up absolute time if you need to
    - when quoting quotes, relative times will be adjusted

# Notes

- Quoting quotes now requires as many calls as there are quoted messages
- All other exisiting functionality will break because 1.0.0a has changed a lot of syntax (https://discordpy.readthedocs.io/en/rewrite/migrating.html#major-model-changes). Someone will need to fix those.

# To-Do

- [x] Migrate all the other functions to `async/await` syntax
- [x] Re-do !misquote and SF frames to 1.0.0a syntax
    - [x] Kevin said ignore SF frames

### CODE REVIEW

File: discord_quote/discord_quote/discord_quote.py [hunk 1]
```diff
@@ -91,125 +92,295 @@ def me(ctx, *text : str):
...
+                              'quote',
+                              ctx.message.channel.name]))
+        else:
+            await bot_quote(ctx, msg_, *reply)
+    except discord.errors.HTTPException as e:
+        log.warning(['msg_not_found', msg_id, ctx.message.author.mention, e])
 
         # Return error if message not found.
-        yield from bot.say(("Quote not found in this channel ('{0}' "
+        await ctx.channel.send(("Quote not found in this channel ('{0}' "
```
Comment (synsypa) [inline discord_quote/discord_quote/discord_quote.py]: do we think its worth being consistent with fstring usage or is the nature of this case that a normal formatted string is more illustrative?

File: discord_quote/discord_quote/discord_quote.py [hunk 2]
```diff
@@ -91,125 +92,295 @@ def me(ctx, *text : str):
...
+            message_time = _author.group(2)
+            log.info(log_msg(['found_original_timestamp', msg_.id, message_time]))
+
+    clean_content = msg_.clean_content
+
+    # Format output message, handling replies and quotes
+    if not reply and not quote:
+        log.info(log_msg(['formatting_quote', 'noreply|noquote']))
+        # Simplest case, just quoting a non-quotebot message, with no reply
+        output = '**{0} [{1}] said:** _via {2}_ ```{3}```'.format(
```
Comment (synsypa) [inline discord_quote/discord_quote/discord_quote.py]: see above

File: discord_quote/discord_quote/discord_quote.py [hunk 3]
```diff
@@ -91,125 +92,295 @@ def me(ctx, *text : str):
...
+        log.info(log_msg(['formatting_quote', 'reply|quote']))
+
+        # Detect Last Response so we can hyperlink
+        _last_response = re.search(
+                "\*\*[A-Za-z0-9]*\s(\[[A-Za-z0-9\s]*\])\sresponded",
+                msg_.content
+        )
+
+        if _last_response:
+            clean_content = clean_content.replace(
```
Comment (synsypa) [inline discord_quote/discord_quote/discord_quote.py]: see again

File: discord_quote/discord_quote/discord_quote.py [hunk 4]
```diff
@@ -91,125 +92,295 @@ def me(ctx, *text : str):
...
+                    _last_response.group(1),
+                    "[{0}({1})]".format(_last_response.group(1), jump_url)
+            )
+
+        # Reply to a quotebot quote with a reply
+        output = '{0}\n**{1} [{2}] responded:** {3}'.format(
```
Comment (synsypa) [inline discord_quote/discord_quote/discord_quote.py]: more

File: discord_quote/discord_quote/discord_quote.py [hunk 5]
```diff
@@ -91,125 +92,295 @@ def me(ctx, *text : str):
...
+                            clean_content,
+                            ctx.message.author.name,
+                            ctx.message.created_at.strftime("%Y-%m-%d %H:%M:%S"),
+                            ' '.join(reply)
+                    )
+    else:
+        log.info(log_msg(['formatting_quote', 'reply|quote']))
+
+        output = (
```
Comment (synsypa) [inline discord_quote/discord_quote/discord_quote.py]: more

File: discord_quote/discord_quote/discord_quote.py [hunk 6]
```diff
@@ -226,35 +397,45 @@ def misquote(ctx , target : discord.User):
...
-        reply = yield from bot.wait_for_message(timeout=60.0,
-                                                author=ctx.message.author,
-                                                check=priv)
+        # Check that this is the right message
+        def pred(m):
+            return (
+                m.author == ctx.message.author 
+                and isinstance(m.channel, discord.DMChannel)
+            )
+        reply = await bot.wait_for('message', check=pred)
```
Comment (synsypa) [inline discord_quote/discord_quote/discord_quote.py]: i think this still takes `timeout = 60`

Comment (zhangchuck) [reply discord_quote/discord_quote/discord_quote.py]: added

**Comment by synsypa** [comment]
lgtm

**Pull request merged**